gcc/rust/checks/errors/borrowck/rust-bir-place.h

```diff
@@ -92,60 +74,192 @@ struct Place
   } path;
   /** Copy trait */
   bool is_copy;
-  /** This place can be moved from safety. */
-  bool is_rvalue;
-  Lifetime lifetime;
+  bool has_drop = false;
   TyTy::BaseType *tyty;
+  FreeRegions regions{{}};
+  std::vector<LoanId> borrowed_by{};
 
+public:
   Place (Kind kind, uint32_t variable_or_field_index, const Path &path,
-	 bool is_copy, bool is_rvalue, const Lifetime &lifetime,
-	 TyTy::BaseType *tyty)
+	 bool is_copy, TyTy::BaseType *tyty)
     : kind (kind), variable_or_field_index (variable_or_field_index),
-      path (path), is_copy (is_copy), is_rvalue (is_rvalue),
-      lifetime (lifetime), tyty (tyty)
+      path (path), is_copy (is_copy), tyty (tyty)
   {}
+
+  // Place can only be stored in PlaceDB and used via reference. Turn all
+  // accidental copies into errors.
+  Place (const Place &) = delete;
+  Place (Place &&) = default;
+
+public:
+  WARN_UNUSED_RESULT bool is_lvalue () const
+  {
+    return kind == VARIABLE || is_path ();
+  }
+
+  WARN_UNUSED_RESULT bool is_rvalue () const { return kind == TEMPORARY; }
+
+  bool is_constant () const { return kind == CONSTANT; }
+
+  WARN_UNUSED_RESULT bool is_var () const
+  {
+    return kind == VARIABLE || kind == TEMPORARY;
+  }
+
+  WARN_UNUSED_RESULT bool is_path () const
+  {
+    return kind == FIELD || kind == INDEX || kind == DEREF;
+  }
+
+  WARN_UNUSED_RESULT TyTy::BaseType *get_fn_return_ty () const
+  {
+    switch (tyty->get_kind ())
+      {
+      case TyTy::FNPTR:
+	return tyty->as<TyTy::FnPtr> ()->get_return_type ();
+      case TyTy::FNDEF:
+	return tyty->as<TyTy::FnType> ()->get_return_type ();
+      default:
+	rust_assert (false);
+      }
+  }
+
+  WARN_UNUSED_RESULT bool is_indirect () const
+  {
+    // TODO: probably incomplete, check other projections
+    switch (tyty->get_kind ())
+      {
+      case TyTy::REF:
+      case TyTy::POINTER:
+	return true;
+      default:
+	return false;
+      }
+  }
+
+  WARN_UNUSED_RESULT bool should_be_moved () const
+  {
+    return kind == TEMPORARY || (!is_copy && kind != CONSTANT);
+  }
+};
+
+using ScopeId = uint32_t;
+
+static constexpr ScopeId INVALID_SCOPE = std::numeric_limits<ScopeId>::max ();
+/** Arguments and return value are in the root scope. */
+static constexpr ScopeId ROOT_SCOPE = 0;
+/** Top-level local variables are in the top-level scope. */
```

Comment:
is this taken from anywhere? rustc's implementation of NLL with polonius?

Reply:
I don't think it is from rustc code, but it is how the MIR dump displays it.